### PR TITLE
8329289: Unify SetupJdkExecutable and SetupJdkLibrary

### PR DESCRIPTION
**Pull request by magicus**
Currently a lot of code is duplicated between SetupJdkExecutable and SetupJdkLibrary. Furthermore, some functionality is still missing from SetupJdkExecutable that is present in SetupJdkLibrary. These functions also have not had their documentation properly updated as they have evolved. This PR will fix all of this.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8329289](https://bugs.openjdk.org/browse/JDK-8329289): Unify SetupJdkExecutable and SetupJdkLibrary (**Bug** - P3)

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**) ⚠️ Review applies to [06050fc7](https://git.openjdk.org/jdk/pull/18537/files/06050fc72d809a38c25fa60d98aeac03b2396a45)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18537/head:pull/18537` \
`$ git checkout pull/18537`

Update a local copy of the PR: \
`$ git checkout pull/18537` \
`$ git pull https://git.openjdk.org/jdk.git pull/18537/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18537`

View PR using the GUI difftool: \
`$ git pr show -t 18537`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18537.diff">https://git.openjdk.org/jdk/pull/18537.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18537#issuecomment-2025812936)

### CODE REVIEW

**Comment by magicus** [comment]
Using COMPARE_BUILD, there is no changes to any executables or production libraries. 

However, for some reason the gtest libjvm.so is 8 bytes larger on linux-x64. I can't really tell what causes this. It might be a problem with reproducible builds, or perhaps that some command line arguments has been reordered. I don't think it matters.

File: make/common/JdkNativeCompilation.gmk [hunk 1]
```diff
@@ -199,11 +218,19 @@ define SetupJdkLibraryBody
...
-    $1_CXXFLAGS := $$(filter-out $$($1_CXXFLAGS_FILTER_OUT), $$(CXXFLAGS_JDKLIB)) $$($1_CXXFLAGS)
+      # Set the default flags first to be able to override
+      $1_CXXFLAGS := $$(filter-out $$($1_CXXFLAGS_FILTER_OUT), $$(CXXFLAGS_JDKEXE)) $$($1_CXXFLAGS)
+    else
+      # Set the default flags first to be able to override
+      $1_CFLAGS := $$(filter-out $$($1_CFLAGS_FILTER_OUT), $$(CFLAGS_JDKLIB)) $$($1_CFLAGS)
+
+      # Set the default flags first to be able to override
+      $1_CXXFLAGS := $$(filter-out $$($1_CXXFLAGS_FILTER_OUT), $$(CXXFLAGS_JDKLIB)) $$($1_CXXFLAGS)
+    endif
```
Comment (erikj79) [inline make/common/JdkNativeCompilation.gmk:233]: I think it makes sense to share all that is actually common between the two existing macros, but for these conditional adding default flags, it's just a big if EXECUTABLE do this, otherwise do that. I think in such cases it makes more sense to keep that logic in the respective specialized macros. The only drawback would be that the new `SetupJdkNativeCompilation` won't be usable on its own, but it's not intended to be anyway.

Comment (magicus) [reply make/common/JdkNativeCompilation.gmk:233]: Actually, I'm intending on calling SetupJdkNativeCompilation directly in one place, from SetupTestFilesCompilation, which already mixes executable and libraries at it's entry point. So I'd rather keep SetupJdkNativeCompilation working.

Also, I plan to follow this up with a PR that splits up the C[XX]FLAGS_JDK[EXE|LIB] into constituent parts and then just select the few that differs between C and C++, LIB and EXE depending on which is chosen. I should perhaps have been clearer about this.

Comment (magicus) [reply make/common/JdkNativeCompilation.gmk:233]: I also see another couple of rounds of refactoring on SetupJdkNativeCompilation and SetupNativeCompilation before they get really into shape. I prefer to do that step by step, so it is possible to follow what changes are being done. If I were to make all changes I'd like and push it in a single PR, it would be very hard to follow how the code has been transformed.

**Review by erikj79 — APPROVED** [review]
